File: frontend/src/utils/atwDocumentParser.ts
export interface AtwDocumentMetadata {
  atwNumber?: string
  issueDate?: string
  expirationDate?: string
  containerNumbers: string[]
  destination?: string
}

function toIsoDate(match: RegExpMatchArray): string | undefined {
  const [, y, m, d] = match
  if (!y || !m || !d) return undefined
  const year = y.length === 2 ? `20${y}` : y
  return `${year}-${m.padStart(2, '0')}-${d.padStart(2, '0')}`
}

export function parseAtwDocumentText(text: string): AtwDocumentMetadata {
  const upper = text.toUpperCase()
  const containerNumbers = [
    ...new Set(
      [...upper.matchAll(/\b([A-Z]{4}\d{7})\b/g)].map((m) => m[1]).filter(Boolean),
    ),
  ]

  const atwMatch =
    upper.match(/\bATW[\s\-#:]*([A-Z0-9\-]{4,24})\b/) ??
    upper.match(/\bAUTHORITY\s+TO\s+WITHDRAW[\s\S]{0,40}?([A-Z0-9\-]{5,24})\b/)

  const issueMatch =
    upper.match(/ISSUE\s*DATE[:\s]+(\d{1,2})[\/\-](\d{1,2})[\/\-](\d{2,4})/) ??
    upper.match(/DATE\s+ISSUED[:\s]+(\d{1,2})[\/\-](\d{1,2})[\/\-](\d{2,4})/)

  const expiryMatch =
    upper.match(/EXPIR(?:Y|ATION)\s*DATE[:\s]+(\d{1,2})[\/\-](\d{1,2})[\/\-](\d{2,4})/) ??
    upper.match(/VALID\s+UNTIL[:\s]+(\d{1,2})[\/\-](\d{1,2})[\/\-](\d{2,4})/)

  const destinationMatch = upper.match(/DESTINATION[:\s]+([A-Z0-9 ,.\-]{4,60})/)

  return {
    atwNumber: atwMatch?.[1]?.replace(/[^A-Z0-9\-]/g, ''),
    issueDate: issueMatch ? toIsoDate(issueMatch) : undefined,
    expirationDate: expiryMatch ? toIsoDate(expiryMatch) : undefined,
    containerNumbers,
    destination: destinationMatch?.[1]?.trim().replace(/\s{2,}/g, ' '),
  }
}

File: android-trucker/app/src/main/java/com/ecms/trucker/util/AtwDocumentParser.kt
package com.ecms.trucker.util

data class AtwDocumentMetadata(
    val atwNumber: String? = null,
    val issueDate: String? = null,
    val expirationDate: String? = null,
    val containerNumbers: List<String> = emptyList(),
    val destination: String? = null,
)

object AtwDocumentParser {
    fun parse(text: String): AtwDocumentMetadata {
        val upper = text.uppercase()
        val containerNumbers = Regex("""\b([A-Z]{4}\d{7})\b""")
            .findAll(upper)
            .map { it.groupValues[1] }
            .distinct()
            .toList()

        val atwMatch = Regex("""\bATW[\s\-#:]*([A-Z0-9\-]{4,24})\b""").find(upper)
            ?: Regex("""\bAUTHORITY\s+TO\s+WITHDRAW[\s\S]{0,40}?([A-Z0-9\-]{5,24})\b""").find(upper)

        val issueMatch = Regex("""ISSUE\s*DATE[:\s]+(\d{1,2})[\/\-](\d{1,2})[\/\-](\d{2,4})""").find(upper)
            ?: Regex("""DATE\s+ISSUED[:\s]+(\d{1,2})[\/\-](\d{1,2})[\/\-](\d{2,4})""").find(upper)

        val expiryMatch = Regex("""EXPIR(?:Y|ATION)\s*DATE[:\s]+(\d{1,2})[\/\-](\d{1,2})[\/\-](\d{2,4})""").find(upper)
            ?: Regex("""VALID\s+UNTIL[:\s]+(\d{1,2})[\/\-](\d{1,2})[\/\-](\d{2,4})""").find(upper)

        val destinationMatch = Regex("""DESTINATION[:\s]+([A-Z0-9 ,.\-]{4,60})""").find(upper)

        return AtwDocumentMetadata(
            atwNumber = atwMatch?.groupValues?.getOrNull(1)?.replace(Regex("""[^A-Z0-9\-]"""), ""),
            issueDate = issueMatch?.let { toIsoDate(it) },
            expirationDate = expiryMatch?.let { toIsoDate(it) },
            containerNumbers = containerNumbers,
            destination = destinationMatch?.groupValues?.getOrNull(1)?.trim()?.replace(Regex("""\s{2,}"""), " "),
        )
    }

    private fun toIsoDate(match: MatchResult): String? {
        val y = match.groupValues.getOrNull(3) ?: return null
        val m = match.groupValues.getOrNull(2) ?: return null
        val d = match.groupValues.getOrNull(1) ?: return null
        val year = if (y.length == 2) "20$y" else y
        return "$year-${m.padStart(2, '0')}-${d.padStart(2, '0')}"
    }
}
